### PR TITLE
fix: improve linting and automation

### PR DESCRIPTION
**Pull request by RobotSail**
In the current version of the training repo, we have a few commands that help us lint
and check to see if there are any issues with types or incorrect imports. However;
there is no easy fix available when these issues are detected.

This commit introduces a few new functionalities:

1. Makefile receives 5 new commands:
 - toml-lint, which lints the pyproject.toml file
 - toml-fmt, which automatically formats the pyproject.toml file
 - check-engine, which checks that the specified container engine ($CENGINE) is installed
 - tests, which runs tests against the source repo
 - fix, which fixes all fixable issues

2. The tox.ini file receives a few additions:
  a. The testenv environment is defined with some preset values. Other targets are able to pull from this
  b. tox -e fix is added - automatically fixes all fixable problems
  c. tox -e tomllint is added - fixes all issues in the pyproject.toml file
  d. a [py] environment is added to specify the base python version

Aside from this, the resulting pyproject.toml file is added after formatting  is applied.

Signed-off-by: Oleg S <97077423+RobotSail@users.noreply.github.com>

### CODE REVIEW

**Comment by nathan-weinberg** [comment]
@mergifyio refresh

File: Makefile [hunk 1]
```diff
@@ -54,3 +71,15 @@ spellcheck-sort: .spellcheck-en-custom.txt ## Sort spellcheck directory
...
+
+
+# TODO: add unit tests in the future
+.PHONY: tests
+tests: check-tox ## Run type checks.
+	tox -e mypy
+
+
+.PHONY: fix
+fix: check-tox ## Fix everything that's fixable by the automaed tooling.
```
Comment (nathan-weinberg) [inline Makefile:83]:
```suggestion
fix: check-tox ## Fix everything that's fixable by the automated tooling.
```

**Review by nathan-weinberg — COMMENTED** [review]
Generally LGTM, not sure why a new lint error is getting picked up? Maybe needs a rebase?

File: Makefile (hunk 1 above)
Comment (RobotSail) [reply Makefile:83]: Thank you sir

**Comment by RobotSail** [comment]
@mergify rebase

**Comment by RobotSail** [comment]
@nathan-weinberg Tests are passing now

**Review by nathan-weinberg — APPROVED** [review]

**Review by JamesKunstle — APPROVED** [review]
looks like a good addition to me

**Pull request merged**